### PR TITLE
Preventing sending Basic Auth headers with "None:None"

### LINKED ISSUE #211: Wrong Authorization header?
I'm having trouble using requests-oauthlib against a flask-oauthlib provider. I'm using the [reference implementation of flask-oauthlib provider](https://github.com/lepture/example-oauth2-server/) and while using the same library for the client works, using requests-oauthlib fails.

What I do is:
```
client_id = 'AHc6492FZUnD6l9kAENvU8rM8AQyEZAOYnLf2JEi'
client_secret = 'NJWNqoG5wh2QQ3aCobnmaHNzcNBRozL60p7S8zGcHwXWOYNEdN'
redirect_uri = 'http://127.0.0.1:8000/authorized'
authorization_base_url = 'http://127.0.0.1:5000/oauth/authorize'
token_url = "http://127.0.0.1:5000/oauth/token"
scope = ['email']
from requests_oauthlib import OAuth2Session
provider = OAuth2Session(client_id,scope=scope,redirect_uri=redirect_uri)
authorization_url, state = provider.authorization_url(authorization_base_url)
print("GO TO: "+authorization_url)

#this prints the auth url:
GO TO: http://127.0.0.1:5000/oauth/authorize?response_type=code&client_id=AHc6492FZUnD6l9kAENvU8rM8AQyEZAOYnLf2JEi&redirect_uri=http%3A%2F%2F127.0.0.1%3A8000%2Fauthorized&scope=email&state=vgta4Exp0v4dPiJdspfIlXOeTgvN3Q

# I go there and authorize the client. I get to go to a redirected url, that I paste here:
url = "http://127.0.0.1:8000/authorized?state=vgta4Exp0v4dPiJdspfIlXOeTgvN3Q&code=r6s8qAAYqJzffFi3iaKl02KgSTT20I"
provider.fetch_token(token_url, client_secret=client_secret, authorization_response=url)
```

This last line raises a `InvalidClientError` exception.

```
InvalidClientError                        Traceback (most recent call last)
<ipython-input-7-5ff386dbf014> in <module>()
      1 url = "http://127.0.0.1:8000/authorized?state=vgta4Exp0v4dPiJdspfIlXOeTgvN3Q&code=r6s8qAAYqJzffFi3iaKl02KgSTT20I"
----> 2 repovizz.fetch_token(token_url, client_secret=client_secret, authorization_response=url)

/home/carles/tmp/example-oauth2-server/local/lib/python2.7/site-packages/requests_oauthlib/oauth2_session.pyc in fetch_token(self, token_url, code, authorization_response, body, auth, username, password, method, timeout, headers, verify, **kwargs)
    230             r = hook(r)
    231 
--> 232         self._client.parse_request_body_response(r.text, scope=self.scope)
    233         self.token = self._client.token
    234         log.debug('Obtained token %s.', self.token)

/home/carles/tmp/example-oauth2-server/local/lib/python2.7/site-packages/oauthlib/oauth2/rfc6749/clients/base.pyc in parse_request_body_response(self, body, scope, **kwargs)
    407         .. _`Section 7.1`: http://tools.ietf.org/html/rfc6749#section-7.1
    408         """
--> 409         self.token = parse_token_response(body, scope=scope)
    410         self._populate_attributes(self.token)
    411         return self.token

/home/carles/tmp/example-oauth2-server/local/lib/python2.7/site-packages/oauthlib/oauth2/rfc6749/parameters.pyc in parse_token_response(body, scope)
    374 
    375     params = OAuth2Token(params, old_scope=scope)
--> 376     validate_token_parameters(params)
    377     return params
    378 

/home/carles/tmp/example-oauth2-server/local/lib/python2.7/site-packages/oauthlib/oauth2/rfc6749/parameters.pyc in validate_token_parameters(params)
    381     """Ensures token precence, token type, expiration and scope in params."""
    382     if 'error' in params:
--> 383         raise_from_error(params.get('error'), params)
    384 
    385     if not 'access_token' in params:

/home/carles/tmp/example-oauth2-server/local/lib/python2.7/site-packages/oauthlib/oauth2/rfc6749/errors.pyc in raise_from_error(error, params)
    269     for _, cls in inspect.getmembers(sys.modules[__name__], inspect.isclass):
    270         if cls.error == error:
--> 271             raise cls(**kwargs)

InvalidClientError: (invalid_client) 
```

It seems that in the header is passed like this: `u\'Authorization\': u\'Basic Tm9uZTpOb25l\'`, which is used by flask-oauthlib to retrieve the client id and secret, by deserializing this using base64 encoding, which results in `"None:None"`. This triggers the HTTP 401 client error response.

Am I doing something wrong here? It seems that requests-oauthib is creating faulty headers, but I am no expert in oauthlib2 and that may be also a wrong/restricted implementation of flask-oauthlib...

Versions:
```
requests-oauthlib==0.6.0
Flask-OAuthlib==0.9.2
```

### PR DESCRIPTION
**Pull request by chaosct**
Closes #211

### CODE REVIEW

File: requests_oauthlib/oauth2_session.py [hunk 1]
```diff
@@ -197,7 +197,8 @@ def fetch_token(self, token_url, code=None, authorization_response=None,
                 redirect_uri=self.redirect_uri, username=username,
                 password=password, **kwargs)
 
-        auth = auth or requests.auth.HTTPBasicAuth(username, password)
+        if not auth and (username):
+            auth = auth or requests.auth.HTTPBasicAuth(username, password)
```
Comment (Lukasa) [inline requests_oauthlib/oauth2_session.py]: Let's tighten this up a little bit if you don't mind. =)

Firstly, we can remove the `auth or` portion of this line: we already know `auth` is false, that's how we got here, so no need to check it again.

Secondly, let's add a line inside this block that asserts that `password` is truthy.

Thirdly and finally, let's remove the parentheses around `username` in the conditional: they're not needed. =)

**Comment by Lukasa** [comment]
Thanks! Some small notes. =)

**Comment by chaosct** [comment]
Should I make those changes? Or are you already on it?

**Comment by Lukasa** [comment]
I'm happy for you to do it. =)

**Comment by chaosct** [comment]
For the password, I guess an empty password is OK? I'll then compare it with `is None`.

And with this code, if username is empty string, it won't use Basic Auth, but I am not sure if empty string is a valid username or password.

**Comment by Lukasa** [comment]
@chaosct I don't believe an empty string is a valid username.

**Comment by chaosct** [comment]
Let me know if more changes are needed

File: requests_oauthlib/oauth2_session.py [hunk 2]
```diff
@@ -197,7 +197,10 @@ def fetch_token(self, token_url, code=None, authorization_response=None,
                 redirect_uri=self.redirect_uri, username=username,
                 password=password, **kwargs)
 
-        auth = auth or requests.auth.HTTPBasicAuth(username, password)
+        if not auth and username:
```
Comment (Lukasa) [inline requests_oauthlib/oauth2_session.py]: Heh, reading this now I find the operator precedence confusing. Let's do `if (not auth) and username` just to indicate what's going on here.

**Comment by Lukasa** [comment]
Beautiful, almost perfect. One tiny change and then it's good to go!

**Comment by chaosct** [comment]
Done!

**Comment by Lukasa** [comment]
\o/ Perfect! Thanks @chaosct!

**Comment by citizenken** [comment]
Can a new release be created to include these changes please? Installing master straight from Github is not ideal.

**Comment by sigmavirus24** [comment]
@citizenken it would appear this results in a new bug so creating a release right now does not seem ideal

**Pull request merged**